samples: ironside_se: Add sample that demonstrates snapshot (#28633)
nrfconnect/sdk-nrf: samples/ironside_se/snapshot_capture_recover/README.rst

.. _ironside_se_snapshot_capture_recover:

> IronSide SE: Snapshot capture and recover

Reviewer: Codeowners file and changelog  to be updated.

Author: @umapraseeda : Can you elaborate on the update to CODEOWNERS?

Is it not covered by the existing

/samples/ironside_se/                     @nrfconnect/ncs-aurora

?

Author: added an entry in the changelog

Reviewer: Yes, you are right, there is an entry:  https://github.com/nrfconnect/sdk-nrf/blob/main/CODEOWNERS#L670.

#########################################

.. contents::
   :local:
   :depth: 2

Overview
********

> This sample demonstates IronSide SE snapshot capture and recovery.

Reviewer 2: ```suggestion
This sample demonstrates the IronSide SE snapshot capture and recovery features.
This feature helps in recovering the content of the MRAM when corruption due to bit flips caused caused by magnetic fields occurs.
```

Author: fixed

It performs a couple of snapshot capture's and snapshot recovery
operations and then goes into an infinite loop where the user may
choose to apply a magnet to corrupt MRAM and observe that the device

> recovers.

Reviewer 2: ```suggestion
The application performs several snapshot capture and snapshot recovery operations.
After completing these operations, it enters an infinite loop.

In this loop, you can apply a magnet to corrupt the MRAM and observe that the device recovers.
```

Author: fixed

> To keep track of which boot we are on we use an IronSide SE NV

Reviewer 2: What's that, exactly?

Author: fixed

counter.

UICR configuration:

* **UICR.LOCK** — Not written by :file:`sysbuild/uicr.conf`. After each boot the sample reads the NV counter, if the value was ``0``, it then calls ``uicr_deploy_lock_contents()`` and cold-reboots.
* **PROTECTEDMEM** — Protects ``cpuapp_boot_partition`` and ``periphconf_partition`` (72 KB), consistent with other IronSide SE samples that place periphconf immediately after the boot partition.
* **UICR.SNAPSHOT.REGIONS** — Region 0 at physical ``0x0E030000`` (72 KB) for boot + periphconf; additional regions in :file:`sysbuild/uicr.conf` cover secure storage and the MRAM area used by the sample stress path.

The application validates the boot report, increments an NV counter, logs the snapshot status, and then drives capture → recovery → cold reboot.
To keep the sample bounded and repeatable, the flow stops once the counter reaches ``SNAPSHOT_MAX_CYCLES`` in :file:`src/main.c`.

> After that, it enters a heartbeat loop and periodically performs high-volume MRAM read sweeps over a large region to help surface possible MRAM corruption.

Reviewer 3: and hence trigger a recovery?

Author: fixed

Requirements
************

> * nRF54H20 DK (application core)

Reviewer 2: ```suggestion
The sample supports the following development kit:

.. table-from-sample-yaml::

```

Author: fixed

> * IronSide SE with snapshot and counter services

Reviewer 2: ```suggestion
The sample also requires your DK to be updated to the latest version of IronSide SE .
For more information, see :ref:`abi_compatibility`
```

Author: fixed

Building and running
********************

.. code-block:: console

   west build -b nrf54h20dk/nrf54h20/cpuapp --sysbuild nrf/samples/ironside_se/snapshot_capture_recover

>    west flash

Reviewer 2: ```suggestion
.. |sample path| replace:: :file:`samples/ironside_se/snapshot_capture_recover`

.. include:: /includes/build_and_run.txt
```

Author: fixed

> To reset persistent state (including counters) before a clean run:

Reviewer 2: ```suggestion
To reset the persistent state, including all counters, before starting a clean execution, run the following command:
```

Author: fixed

.. code-block:: console

>    west flash --recover

Reviewer 2: What about testing and dependencies?

Author: What about testing and dependencies?

:)

Author: Added a section on testing and another on which dependencies the sample has.

.. _ironside_se_snapshot_capture_recover:

IronSide SE: Snapshot capture and recover
#########################################

.. contents::
   :local:
   :depth: 2

> Overview

Reviewer 4: The Overview section comes after the Requirements section, see the template for sample: https://ncsdoc.z6.web.core.windows.net/ncs/latest/nrf/templates/sample_README.html

> Overview
> ********
> 

Reviewer 2: ```suggestion
```

> This sample demonstrates the IronSide SE snapshot capture and recovery features.

Reviewer 4: Move lines 13 and 14 before the Requirements section to provide an introduction to the sample.
```suggestion
The Snapshot capture and recover sample demonstrates the |ISE| snapshot capture and recovery features.
```

This feature helps in recovering the content of the MRAM when corruption due to bit flips caused by magnetic fields occurs.

> The application performs several snapshot capture and snapshot recovery operations.

Reviewer 4: Application or Sample?

After completing these operations, it enters an infinite loop.

In this loop, you can apply a magnet to corrupt the MRAM and observe that the device recovers.

To keep track of which boot the sample is on, it uses an |ISE| NV counter (:ref:`ug_nrf54h20_ironside_se_counter_service`).

UICR configuration:

> * **UICR.LOCK** - Not written by :file:`sysbuild/uicr.conf`. After each boot the sample reads the NV counter; if the value was ``0``, it then calls ``uicr_deploy_lock_contents()`` and cold-reboots.

Reviewer 4: ```suggestion
* **UICR.LOCK** - Not written by :file:`sysbuild/uicr.conf`.
  After each boot the sample reads the NV counter; if the value was ``0``, it then calls ``uicr_deploy_lock_contents()`` and cold-reboots.
```

> * **PROTECTEDMEM** - Protects ``cpuapp_boot_partition`` and ``periphconf_partition`` (72 KB), consistent with other IronSide SE samples that place periphconf immediately after the boot partition.

Reviewer 4: ```suggestion
* **PROTECTEDMEM** - Protects ``cpuapp_boot_partition`` and ``periphconf_partition`` (72 KB), consistent with other |ISE| samples that place periphconf immediately after the boot partition.
```

Author: fixed

* **UICR.SNAPSHOT.REGIONS** - Region 0 at physical ``0x0E030000`` (72 KB) for boot + periphconf; additional regions in :file:`sysbuild/uicr.conf` cover secure storage and the MRAM area used by the sample stress path.

The application validates the boot report, increments an NV counter, logs the snapshot status, and then drives capture → recovery → cold reboot.
To keep the sample bounded and repeatable, the flow stops once the counter reaches ``SNAPSHOT_MAX_CYCLES`` in :file:`src/main.c`.
After that, it enters a heartbeat loop and periodically performs high-volume MRAM read sweeps over a large region.

> If |ISE| detects corruption in this region it will trigger snapshot recovery.

Reviewer 2: ```suggestion
```
See comment below

Requirements
************

The sample supports the following development kit:

.. table-from-sample-yaml::

> The sample also requires your DK to be updated to the latest version of |ISE|.

Reviewer 2: ```suggestion
The sample also requires your development kit to be on the latest version of |ISE|.
```

For more information, see :ref:`abi_compatibility`.

> 

Reviewer 2: ```suggestion

Overview
********

The sample performs multiple snapshot capture and recovery operations, followed by an infinite loop of reboot cycles.

In each cycle, the sample validates the boot report, increments an |ISE| NV counter (:ref:`ug_nrf54h20_ironside_se_counter_service`), logs the snapshot status, and executes a *capture → recovery → cold* reboot sequence.

To keep execution bounded and repeatable, the sequence stops when the counter reaches ``SNAPSHOT_MAX_CYCLES`` in :file:`src/main.c`.  
After reaching this limit, the sample enters a heartbeat loop.

During the heartbeat loop, the sample periodically performs high-volume MRAM read sweeps over a large region.  
If |ISE| detects corruption in this region, it triggers snapshot recovery.

You can apply a magnet during the heartbeat loop to intentionally corrupt the MRAM and observe that the device recovers.

UICR configuration
==================

The sample configures the following UICR settings:

* UICR.LOCK – Not written by :file:`sysbuild/uicr.conf`.  
  After each boot, the sample reads the NV counter.  
  If the value is ``0``, it calls ``uicr_deploy_lock_contents()`` and performs a cold reboot.

* PROTECTEDMEM – Protects ``cpuapp_boot_partition`` and ``periphconf_partition`` (72 KB).  
  This layout is consistent with other |ISE| samples, where ``periphconf`` is placed immediately after the boot partition.

* UICR.SNAPSHOT.REGIONS – Defines snapshot regions.  
  Region 0 is located at physical address ``0x0E030000`` (72 KB) and covers the boot and ``periphconf`` partitions.  
  Additional regions defined in :file:`sysbuild/uicr.conf` cover secure storage and the MRAM area used by the sample stress path.

```

Moved here and edited

Author: this is a big diff, so not sure if I got everything.

Requirements
************

The sample supports the following development kit:

.. table-from-sample-yaml::

The sample also requires your development kit to be on the latest version of |ISE|.
For more information, see :ref:`abi_compatibility`.

Overview
********

The sample performs multiple snapshot capture and recovery operations, followed by an infinite loop of reboot cycles.

In each cycle, the sample validates the boot report, increments an |ISE| NV counter (:ref:`ug_nrf54h20_ironside_se_counter_service`), logs the snapshot status, and executes a *capture → recovery → cold* reboot sequence.

To keep execution bounded and repeatable, the sequence stops when the counter reaches ``SNAPSHOT_MAX_CYCLES`` in :file:`src/main.c`.
After reaching this limit, the sample enters a heartbeat loop.

During the heartbeat loop, the sample periodically performs high-volume MRAM read sweeps over a large region.
If |ISE| detects corruption in this region, it triggers snapshot recovery.

You can apply a magnet during the heartbeat loop to intentionally corrupt the MRAM and observe that the device recovers.

UICR configuration
==================

The sample configures the following UICR settings:

> * **UICR.LOCK** - Not written by :file:`sysbuild/uicr.conf`.

Reviewer 2: ```suggestion
* UICR.LOCK - Not written by :file:`sysbuild/uicr.conf`.
```

  After each boot, the sample reads the NV counter.
  If the value is ``0``, it calls ``uicr_deploy_lock_contents()`` and performs a cold reboot.

> * **PROTECTEDMEM** - Protects ``cpuapp_boot_partition`` and ``periphconf_partition`` (72 KB).

Reviewer 2: ```suggestion
* PROTECTEDMEM - Protects ``cpuapp_boot_partition`` and ``periphconf_partition`` (72 KB).
```

  This layout is consistent with other |ISE| samples, where ``periphconf`` is placed immediately after the boot partition.

> * **UICR.SNAPSHOT.REGIONS** - Defines snapshot regions.

Reviewer 2: ```suggestion
* UICR.SNAPSHOT.REGIONS - Defines snapshot regions.
```
